domain/secret/service/consume.go

```diff
@@ -12,37 +12,57 @@ import (
+// If the secret does not exist, an error satisfying [secreterrors.SecretNotFound] is returned.
+// If there's not currently a consumer record for the secret, the latest revision is still returned,
+// along with an error satisfying [secreterrors.SecretConsumerNotFound].
+func (s *SecretService) GetSecretConsumerAndLatest(ctx context.Context, uri *secrets.URI, unitName string) (*secrets.SecretConsumerMetadata, int, error) {
+	consumerMetadata, latestRevision, err := s.st.GetSecretConsumer(ctx, uri, unitName)
+
+	if err != nil {
+		return nil, latestRevision, errors.Trace(err)
+	}
+	if consumerMetadata.Label != "" {
+		return consumerMetadata, latestRevision, nil
+	}
+	// We allow units to access the application owned secrets using the application owner label,
+	// so we copy the owner label to consumer metadata.
+	md, err := s.getAppOwnedOrUnitOwnedSecretMetadata(ctx, uri, unitName, "")
+	if errors.Is(err, secreterrors.SecretNotFound) {
+		// The secret is owned by a different application.
+		return consumerMetadata, latestRevision, nil
+	}
+	if err != nil {
+		return nil, 0, errors.Annotatef(err, "cannot get secret metadata for %q", uri)
+	}
+	consumerMetadata.Label = md.Label
+	return consumerMetadata, latestRevision, nil
+}
 
-	*/
+// GetSecretConsumer returns the secret consumer info for the specified unit and secret.
+// If the unit does not exist, an error satisfying [uniterrors.NotFound] is returned.
+// If the secret does not exist, an error satisfying [secreterrors.SecretNotFound] is returned.
+// If there's not currently a consumer record for the secret, an error satisfying [secreterrors.SecretConsumerNotFound]
+// is returned.
+func (s *SecretService) GetSecretConsumer(ctx context.Context, uri *secrets.URI, unitName string) (*secrets.SecretConsumerMetadata, error) {
+	result, _, err := s.GetSecretConsumerAndLatest(ctx, uri, unitName)
+	return result, err
 }
 
+// SaveSecretConsumer saves the consumer metadata for the given secret and unit.
+// If the unit does not exist, an error satisfying [uniterrors.NotFound] is returned.
+// If the secret does not exist, an error satisfying [secreterrors.SecretNotFound] is returned.
 func (s *SecretService) SaveSecretConsumer(ctx context.Context, uri *secrets.URI, unitName string, md *secrets.SecretConsumerMetadata) error {
-	return nil
+	return s.st.SaveSecretConsumer(ctx, uri, unitName, md)
 }
 
 func (s *SecretService) GetSecretRemoteConsumer(ctx context.Context, uri *secrets.URI, unitName string) (*secrets.SecretConsumerMetadata, error) {
 	return nil, secreterrors.SecretConsumerNotFound
 }
 
-func (s *SecretService) SaveSecretRemoteConsumer(ctx context.Context, uri *secrets.URI, unitName string, md *secrets.SecretConsumerMetadata) error {
+func (s *SecretService) SaveSecretRemoteConsumer(ctx context.Context, uri *secrets.URI, latestRevision int, unitName string, md *secrets.SecretConsumerMetadata) error {
 	return nil
```

Comment:
NotImplemented?

Reply:
I had this initially but lots of stuff broke. Less breakage to return nil.